server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultConnector.kt

```diff
@@ -254,25 +257,61 @@ class VaultConnector(dispatcher: EventDispatcher<BuildServerListener>, private v
         }
 
         private fun performLogin(template: RestTemplate, settings: VaultFeatureSettings, extractor: (VaultResponse) -> Pair<String, String>): Pair<String, String> {
-            val appRoleAuth = settings.auth as Auth.AppRoleAuthServer
+            when (val auth = settings.auth) {
+                is Auth.AppRoleAuthServer -> {
+                    val options = AppRoleAuthenticationOptions.builder()
+                        .path(auth.getNormalizedEndpoint())
+                        .roleId(auth.roleId)
+                        .secretId(auth.secretId)
+                        .build()
+
+                    val body = getAppRoleLogin(auth)
+                    val path = "auth/${options.path}/login"
+
+                    return performLoginRequest(template, path, body, extractor) { it.replace(auth.secretId, "*******") }
```

Comment:
all three `performLoginRequest` invocations are almost identical except last argument which could be just string which should be replaced, not closure.

Reply:
fixed